Report/final_report.tex

```diff
@@ -277,33 +293,69 @@
+        Retrospectively, because we learned many valuable knowledge from this challenge and now we know how to tackle it, the progress bar is not very difficult. However, although the progress bar itself is trivial, the processes involved in overcoming the challenge are pivotal to us. 
 
-    \subsection{Next steps}
+        \newpage
 
-    We will be carrying the skills we developed in this project into other projects in the future. We have now acquired the ability to download, process, and hash data. These abilities would be necessary to create any automated application in the future.
+        \item The other challenge is about optimizations.
+        
+        In our plot, we implemented a cross-hair that centered on the closet point in the line from the cursor. We successfully implemented it with binary search and many data transformations and calculations. However, the frame rate per second (FPS) was very low when moving the cross-hair.
 
-    We also acquired one of the most important skills in the Computer Science Industry, which is to design a Graphical User Interface. This is the ultimate version of "Designer to User" contract, which provides a very "Easy to use" interface so that users without any knowledge in Computer Science can also use this application to their advantage.
+        To investigate the root of the problem, we analyzed the running time of our algorithms but it turned out that the running time was good ($\bigO(n)$). 
+        
+        Then, we used \verb|time| module and divided the codes into many piece of codes to calculate the real times each piece took. Finally, we found that the root of the evil is actually the \verb|matplotlib#Canvas.draw()| function call. But here comes the issues, we cannot modify the code from \verb|matplotlib| to make it faster.
+        
+        After many researches, we found out that we could solve this issue with a relatively advanced method: blit. Since we already had many experiences from the previous challenge, by reading through the documentation, experimenting with many small examples, and updating the whole implementation of \verb|PlotCanvas|, we successfully implement blit to fix the issue.
+        
+        We firstly draw the figure, excluding the cross-hair, and save the figure background to cache it. Then, every time the cursor moves, we only draw the cross-hair and restore the figure from cache. In this way, the previous cross-hair disappears because the figure background does not contain the cross-hair - we draw the figure before the cross-hair, and we don't need to redraw the background frequently.
 
-    This project will be held public on GitHub repository for anyone to fork and create an updated version of the application. We have included docstrings and comments everywhere around our code for anyone that wishes to fork our Repository.
+        This largely improve the FPS and now the cross-hair is very smooth.
```

Comment:
> This largely improve the FPS and now the cross-hair is very smooth.

This largely improved the frame rate and now the cross-hair is rendered very smoothly.